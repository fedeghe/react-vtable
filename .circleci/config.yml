version: 2
jobs:
  test:
    docker:
      - image: circleci/node:12.16
    steps:
      - checkout
      - run: echo "Running tests"
      - run: yarn
      - run: yarn test
  build:
    docker:
      - image: circleci/node:12.16
    steps:
      - checkout
      - run: echo "build project"
      - run: yarn
      - run: yarn build

workflows:
  version: 2
    test_build:
      jobs:
        - test
        - build:
          requires:
            - test