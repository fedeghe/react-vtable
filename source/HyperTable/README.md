# caption

css
caption-side: top, bottom
text-align: left, center, right

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes

# col

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- span

# colgroup

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- span

# table

# tbody

# td
attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- colspan
- headers
- rowspan

# tfoot

# th
attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- abbr
- colspan
- headers
- rowspan
- scope

# thead
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes

# tr
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes





