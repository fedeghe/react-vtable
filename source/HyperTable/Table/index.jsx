import React, { useCallback, useContext } from 'react'

import Filler from './Filler'
import NoData from './NoData'
import Header from './Header'
import Footer from './Footer'
import LeftMost from './LeftMost'
import RightMost from './RightMost'
import TableContext from './../Context'

import { debounce } from './../utils'

import useStyles from './style.js'


const Table = () => {

    const {state, dispatch} = useContext(TableContext),
        {
            columns,
            rows,
            width, height,
            activeRow, activeCol,
            leftMost, rightMost,
        
            preHeaderHeight, postFooterHeight,
            headerHeight, footerHeight,

            PreHeader, PostFooter,

            crossHighlight,
            columnHighlight,
            rowHighlight,
            cellHightlight,
            cellClick,
            cellEnter,
            cellLeave,
            virtual: {
                headerFillerHeight,
                footerFillerHeight,
                rowHeight,
                from, to
            }
        } = state,
        classes = useStyles({
            width, height,
            preHeaderHeight: PreHeader ? preHeaderHeight : 0,
            postFooterHeight: PostFooter ? postFooterHeight : 0,
            headerHeight,
            footerHeight,
            rowHeight
        }),
        virtualColspan = columns.length + !!leftMost + !!rightMost,
        onScroll = useCallback(debounce(e => {
            e.preventDefault()
            e.stopPropagation()
            dispatch({
                type:'scroll',
                payload: e.nativeEvent.target.scrollTop
            })
        }, 20), []);
    
    return (
        <div
            className={classes.TableContainer}
            onScroll={onScroll}
        >{
            rows.length
            ? (
                <table className={classes.Table}>
                    <Header/>
                    <tbody>
                        <Filler {...{height: headerFillerHeight, virtualColspan}}/>
                        {rows.map((row, i) => (
                            <tr
                                className={`${activeRow === row._ID ? (crossHighlight || rowHighlight || "") : ''}`}
                                key={row._ID}
                            >   
                                <LeftMost cls={`${classes.TbodyThLeftMost} ${classes.Th} ${activeRow === row._ID ? (crossHighlight || rowHighlight) : ''}`} opts={{row, i: i +from, from, to}}/>
                                {columns.map((col, j) => {
                                    let cnt = row[col.key] || 'nothing'
                                    if (col.wrap && typeof col.wrap === 'function') {
                                        cnt = col.wrap(cnt)
                                    }
                                    return (
                                        <td
                                            key={`cell_${row._ID}_${j}`}
                                            onClick={e => {
                                                cellClick.call(e, e, {row, col})
                                            }}
                                            onMouseEnter={e => {
                                                cellEnter.call(e, e, {row, col, rowIndex: from + i, colIndex: j})
                                                dispatch({
                                                    type: 'cellHover',
                                                    payload: {
                                                        row,
                                                        col,
                                                        rowIndex: from + i,
                                                        colIndex: j
                                                    }
                                                });
                                            }}
                                            onMouseLeave={e => {
                                                cellLeave.call(e, e, {row, col,rowIndex: from + i, colIndex: j});
                                                dispatch({type: 'cellOut'});
                                            }}
                                            className={`${classes.Td} ${activeCol === col.key ? (crossHighlight || columnHighlight) : ''} ${(cellHightlight && activeRow === row._ID && activeCol === col.key) ? cellHightlight : ''}`}>
                                            <div className={classes.Cell}>
                                                {cnt}
                                            </div>
                                        </td>
                                    )
                                })}
                                <RightMost cls={`${classes.TbodyThRightMost} ${classes.Th} ${activeRow === row._ID ? (crossHighlight || rowHighlight) : ''}`} opts={{row, i: i + from}}/>
                            </tr>
                        ))}
                        <Filler {...{height: footerFillerHeight, virtualColspan}}/>
                    </tbody>
                    <Footer/>
                </table>
            )
            : <NoData/>
        }
        </div>
    );
}

export default Table