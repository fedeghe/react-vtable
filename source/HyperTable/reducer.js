let count = 0;
const prefix= 'HYT_';
const uniqueID = {
    toString: () => {
        count += 1;
        return prefix + count;
    }
}

let originalData = null

const reducer = (oldState, action) => {
    const { payload, type } = action
    switch (type) {
        case 'sortBy':
            return  {
                ...oldState,
                rows: oldState.rows.sort(payload.sort),
                sorting: {
                    field: payload.field,
                    versus: payload.versus
                },
            }
        case 'unsortBy':
            return  {
                ...oldState,
                rows: [...originalData.reduce((acc, row) => {
                    let canStay = true
                    const {filters} = oldState
                    for (let f in filters) {
                        if (!(`${row[f]}`.includes(`${filters[f]}`))) {
                            canStay = false
                        }
                    }
                    if (canStay) acc.push(row)
                    return acc
                }, [])],
                sorting: {
                    field: null,
                    versus: null
                },
            }
        case 'filter': 
            const newFilters = {
                ...oldState.filters,
                [payload.field]: payload.value
            }
            if (payload.value.length === 0) {
                newFilters[payload.filter] = null;
                delete newFilters[payload.filter]
            }
            // apply all
            return {
                ...oldState,
                filters: newFilters,
                rows: [...originalData.reduce((acc, row) => {
                    let canStay = true
                    for (let f in newFilters) {
                        if (!(`${row[f]}`.includes(`${newFilters[f]}`))) {
                            canStay = false
                        }
                    }
                    if (canStay) acc.push(row)
                    return acc
                }, [])]
            }
        case 'cellHover': 
            return  {
                ...oldState,
                activeCol: payload.col.key,
                activeRow: payload.row._ID
            }
        case 'cellOut': 
            return  {
                ...oldState,
                activeCol: null,
                activeRow: null,
            }
        default:
            return oldState
    }
}

const init = data => {
    originalData = data.map(row => ({_ID: `${uniqueID}`, ...row}))
    return {
        rows: originalData,
        total: originalData.length,
        filters: {},
        sorting: {
            field: null,
            versus: 0
        },
        activeRow: null,
        activeCcol: null
    }
}


export default () => ({
    reducer,
    init
})