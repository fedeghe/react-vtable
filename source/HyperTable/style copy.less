.TableWrapper{
    overflow-y: hidden;
}
.Table {
    display:none;
    border-collapse: collapse;
    background-color: white;
    // display: block;
    overflow: hidden;
    // white-space: nowrap;
    th {
        text-align: left;
    }
    thead, tfoot, tbody tr {
        display: table;
        width: 100%;
        table-layout: fixed;
    }
    tr {
        min-height:1.2em;
    }
    tbody {
        display: block;
        overflow-y: scroll;
        table-layout: fixed;
        overflow-x: hidden; // prevent safari scroll
    }
    .tablebody{
        overflow-y:scroll;
    }
    .tablecell{
        min-width: 30px
    }
    .tableheaderwrapper {
        display:flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
    }
    .tableheaderoptions {
        display:flex;
        flex-direction: row;
        align-items:center;
        input {
            height:1.8em;
            line-height: 1.8em;
            outline:none;
        }
    }
    .tableheadercellsort {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        cursor: pointer;
        font-size: 0.5em;
        margin-left:10px;
        line-height:10px;
    }
    .sortActive {
        color: black
    }
    .sortInactive {
        color: white
    }
    .pointer {
        cursor: pointer;
    }
}
