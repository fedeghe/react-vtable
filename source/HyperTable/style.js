import { createUseStyles } from "react-jss";

export default createUseStyles(() => ({
    Wrapper : ({width, height}) => ({
        height,
        width,
        backgroundColor: 'white',
        overflow: 'hidden',
        position:'relative'
    }),
}));