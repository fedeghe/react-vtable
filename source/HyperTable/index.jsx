import React, { useReducer, useEffect, useRef} from 'react'

import reducerFactory from './reducer'
import { replaceall } from './utils'

import './style.less'

const HyperTable = ({
    columns,
    data,
    captionTop,
    captionBottom,
    height, width,
    rowVerticalAlign,
    crossHighlight,
    columnHighlight,
    rowHighlight,
    cellHightlight,
    noFilterData,
    columnClick = () => { },
    columnEnter = () => { },
    columnLeave = () => { },
    rowClick = () => { },
    rowEnter = () => { },
    rowLeave = () => { },
    cellClick = () => { },
    cellEnter = () => { },
    cellLeave = () => { },
}) => {
    const { reducer, init } = reducerFactory()
    const [state, dispatch] = useReducer(reducer, init(data))
    const {total, rows, activeRow, activeCol, filters, sorting : {
        field: sortingField,
        versus: sortingVersus
    } = {}} = state
    const table = useRef(null);

    

    useEffect(() => {
        table.current.style.display = 'table';
        return () => {
            table.current.style.display = 'none'
        }
    }, [])

    return <div className="TableWrapper">
        <table ref={table} style={{ width: width }} className="Table" border="0" cellSpacing="0" >
            {captionTop && (
                <caption
                    style={{ 'captionSide': 'top'}}
                    className={`TableCaption ${captionTop.className}`}
                >{
                    'component' in captionTop
                    ? captionTop.component({
                        count: rows.length,
                        total, filters,sorting
                    })
                    : replaceall(captionTop.text, {COUNT: rows.length, TOTAL: total})
                }</caption>
            )}
            {captionBottom && (
                <caption
                    style={{ 'captionSide': 'bottom' }}
                    className={`TableCaption ${captionBottom.className}`}
                >{
                    'component' in captionBottom
                    ? captionBottom.component({
                        count: rows.length,
                        total, filters, sorting
                    })
                    : replaceall(captionBottom.text, {COUNT: rows.length, TOTAL: total})
                }</caption>
            )}
            <thead className="TableHeader ">
                <tr>{
                    columns.map((col, k) => (
                        <th key={`h${k}`}
                            className={`TableHeaderCell tablecell`}
                            style={col.width && { width: col.width }}
                        >
                            <div className="tableheaderwrapper">
                                <div>{
                                    'headerComponent' in col
                                        ? col.headerComponent(col, 'key')
                                        : col.headerLabel || col.key
                                }</div>
                                <div className="tableheaderoptions">{
                                    col.filter && col.filter(value => dispatch({
                                        type: 'filter',
                                        payload: {
                                            field: col.key,
                                            value
                                        }
                                    }))
                                }
                                {
                                    col.sorting && (
                                        <div className="tableheadercellsort">
                                            <div className={(sortingField === col.key && sortingVersus === 1) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'sortBy',
                                                    payload: {
                                                        sort: col.sorting.sort(1),
                                                        field: col.key,
                                                        versus: 1
                                                    }
                                                })
                                            }>▲</div>
                                            <div className={(sortingField === null) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'unsortBy',
                                                })
                                            }>&bull;</div>
                                            <div className={(sortingField === col.key && sortingVersus === -1) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'sortBy',
                                                    payload: {
                                                        sort: col.sorting.sort(-1),
                                                        field: col.key,
                                                        versus: -1
                                                    }
                                                })
                                            }>▼</div>
                                        </div>
                                    )
                                }</div>
                            </div>
                        </th>
                    ))
                }</tr>
            </thead>

            <tbody className="TableBody tablebody" style={{ maxHeight: height }}>  
                {rows.length ? rows.map((row, i) => (
                    <tr key={`r${i}`}
                        className={`TableRow ${activeRow === row._ID ? (crossHighlight || rowHighlight) : ''}`}
                        style={{ verticalAlign: rowVerticalAlign || 'top' }}
                        onClick={e => {
                            rowClick.call(e, e, row)
                        }}
                        onMouseEnter={e => {
                            rowEnter.call(e, e, row)
                        }}
                        onMouseLeave={e => {
                            rowLeave.call(e, e, row)
                        }}
                    >{
                            columns.map((col, j) => (
                                <td key={`r${i}c${j}`}
                                    align={col.align || 'left'}
                                    className={`tablecell TableCell ${activeCol === col.key ? (crossHighlight || columnHighlight) : ''} ${(cellHightlight && activeRow === row._ID && activeCol === col.key) ? cellHightlight : ''}`}
                                    style={col.width && { width: col.width }}
                                    onClick={e => {
                                        col.onClick && col.onClick.call(e, e, col, row);
                                        cellClick.call(e, e, col, row);
                                        columnClick.call(e, e, col, row);
                                    }}
                                    onMouseEnter={e => {
                                        dispatch({
                                            type: 'cellHover',
                                            payload: {
                                                row,
                                                col
                                            }
                                        });
                                        cellEnter.call(e, e, row);
                                        columnEnter.call(e, e, col);
                                    }}
                                    onMouseLeave={e => {
                                        dispatch({
                                            type: 'cellOut',
                                            payload: {
                                                row,
                                                col
                                            }
                                        });
                                        cellLeave.call(e, e, row);
                                        columnLeave.call(e, e, col);
                                    }}
                                >{
                                        'component' in col
                                            ? col.component(row, col.key)
                                            : row[col.key] || 'none'
                                }</td>
                            ))
                        }</tr>
                )) : <tr
                    className='TableRow'
                    style={{ verticalAlign: 'middle', textAlign:'center' }}
                ><td className='tablecell TableCell' span={columns.length}>{noFilterData}</td></tr>}
                
            </tbody>



            {columns.some(column => 'footerComponent' in column || 'footerLabel' in column) &&
                <tfoot className="TableFooter">
                    <tr>{
                        columns.map((col, k) => (
                            <th key={`h${k}`}
                                className="TableFooterCell"
                                style={col.width && { width: col.width }}
                            >{
                                    'footerComponent' in col
                                        ? col.footerComponent(col, 'key')
                                        : col.footerLabel
                                }</th>
                        ))
                    }</tr>
                </tfoot>
            }
        </table>
        { /* <pre>{JSON.stringify(state, null, 2)}</pre> */}
    </div>

}

export default HyperTable