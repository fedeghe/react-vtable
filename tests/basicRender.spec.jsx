
/**
 * @jest-environment jsdom
 */
import React from "react";
import {
    render,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { configure } from "@testing-library/dom";

import HyperTable from "../dist/index.js";
import config from "./configs/basic";

configure({
    testIdAttribute: "data-uie",
});

describe("HyperTable - basic", () => {
    it("should render as expected", () => {
        const { container } = render(
            <HyperTable config={config} />
        );
        // check all uie
        expect(container).toMatchSnapshot();
    });
});
