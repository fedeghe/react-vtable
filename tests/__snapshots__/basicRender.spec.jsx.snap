// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HyperTable - basic should render as expected 1`] = `
<div>
  <div
    class="Wrapper-0-2-1 Wrapper-d0-0-2-2 "
  >
    <div
      class="TableContainer-0-2-7 TableContainer-d0-0-2-9"
      tabindex="0"
    >
      <table
        class="Table-0-2-8"
      >
        <tbody>
          <tr
            class="Tr-0-2-22 Tr-d0-0-2-24"
          >
            <td
              class="Td-0-2-23 Td-d1-0-2-25"
              colspan="3"
            />
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d0-0-2-28"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-0"
                >
                  1
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d1-0-2-29"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-0-1"
                >
                  ID0
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d2-0-2-30"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-2"
                >
                  a random, string 0
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d3-0-2-31"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-0"
                >
                  2
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d4-0-2-32"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-1-1"
                >
                  ID1
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d5-0-2-33"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-2"
                >
                  a random, string 1
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d6-0-2-34"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-0"
                >
                  3
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d7-0-2-35"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-2-1"
                >
                  ID2
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d8-0-2-36"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-2"
                >
                  a random, string 2
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d9-0-2-37"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-0"
                >
                  4
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d10-0-2-38"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-3-1"
                >
                  ID3
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d11-0-2-39"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-2"
                >
                  a random, string 3
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d12-0-2-40"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-0"
                >
                  5
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d13-0-2-41"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-4-1"
                >
                  ID4
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d14-0-2-42"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-2"
                >
                  a random, string 4
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d15-0-2-43"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-0"
                >
                  6
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d16-0-2-44"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-5-1"
                >
                  ID5
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d17-0-2-45"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-2"
                >
                  a random, string 5
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d18-0-2-46"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-0"
                >
                  7
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d19-0-2-47"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-6-1"
                >
                  ID6
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d20-0-2-48"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-2"
                >
                  a random, string 6
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d21-0-2-49"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-0"
                >
                  8
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d22-0-2-50"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-7-1"
                >
                  ID7
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d23-0-2-51"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-2"
                >
                  a random, string 7
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d24-0-2-52"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-0"
                >
                  9
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d25-0-2-53"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-8-1"
                >
                  ID8
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d26-0-2-54"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-2"
                >
                  a random, string 8
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d27-0-2-55"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-0"
                >
                  10
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d28-0-2-56"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-9-1"
                >
                  ID9
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d29-0-2-57"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-2"
                >
                  a random, string 9
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d30-0-2-58"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-0"
                >
                  11
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d31-0-2-59"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-10-1"
                >
                  ID10
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d32-0-2-60"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-2"
                >
                  a random, string 10
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d33-0-2-61"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-0"
                >
                  12
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d34-0-2-62"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-11-1"
                >
                  ID11
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d35-0-2-63"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-2"
                >
                  a random, string 11
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d36-0-2-64"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-12-0"
                >
                  13
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d37-0-2-65"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-12-1"
                >
                  ID12
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d38-0-2-66"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-12-2"
                >
                  a random, string 12
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d39-0-2-67"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-13-0"
                >
                  14
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d40-0-2-68"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-13-1"
                >
                  ID13
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d41-0-2-69"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-13-2"
                >
                  a random, string 13
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=" "
            style="padding: 0px;"
          >
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d42-0-2-70"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-14-0"
                >
                  15
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d43-0-2-71"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-14-1"
                >
                  ID14
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-17   "
              style="padding: 0px;"
            >
              <div
                class="Cell-0-2-26 Cell-d44-0-2-72"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-14-2"
                >
                  a random, string 14
                </div>
              </div>
            </td>
          </tr>
          <tr
            class="Tr-0-2-22 Tr-d2-0-2-73"
          >
            <td
              class="Td-0-2-23 Td-d3-0-2-74"
              colspan="3"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
